p2p/discovery.py

```diff
@@ -756,6 +835,53 @@ def get_nodes_to_connect(self, count: int) -> Iterator[kademlia.Node]:
         yield from super().get_nodes_to_connect(num_nodes_needed)
 
 
+class DiscoveryByTopicProtocol(DiscoveryProtocol):
+    """Experimental discovery implementation that uses topic_query to find compatible nodes"""
+    use_v5 = True
+    _concurrent_topic_nodes_requests = 10
+
+    def __init__(self,
+                 topic: bytes,
+                 privkey: datatypes.PrivateKey,
+                 address: kademlia.Address,
+                 bootstrap_nodes: Tuple[kademlia.Node, ...],
+                 cancel_token: CancelToken) -> None:
+        super().__init__(privkey, address, bootstrap_nodes, cancel_token)
+        self.topic = topic
+
+    def get_nodes_to_connect(self, count: int) -> Iterator[kademlia.Node]:
+        topic_nodes = self.topic_table.get_nodes(self.topic)
+        for node in topic_nodes:
+            yield node
+
+        num_nodes_needed = max(0, count - len(topic_nodes))
+        yield from super().get_nodes_to_connect(num_nodes_needed)
+
+    async def lookup_random(self) -> List[kademlia.Node]:
+        query_nodes = list(self.routing.get_random_nodes(self._concurrent_topic_nodes_requests))
+        expected_echoes = []
+        for n in query_nodes:
+            echo = self.send_topic_query(n, self.topic)
+            expected_echoes.append((n, echo))
```

Comment:
I think this might be both more efficient and maybe easier to read as a comprehension.

expected_echos = tuple(
    (n, self.send_topic_query(n, self.topic)),
    for n in query_nodes
)
```

The inlining of the `self.send_topic_query` call is the only part that is a little questionable in terms of readability for me but I think it should e fine.